Unable to forward frame errors followed by panic

```
time="2015-11-04T21:22:55Z" level=warning msg="tchannel: Unable to forward frame ID 6581 type messageTypeError length 91 to peer: timeout" connID=1461 localPeer=10.28.43.68:47651 remotePeer=10.160.4.29:21305 service=moe
time="2015-11-04T21:23:27Z" level=warning msg="tchannel: Unable to forward frame ID 5 type messageTypeError length 91 to peer: timeout" connID=19 localPeer=10.28.43.68:49743 remotePeer=10.28.27.5:55590 service=moe
```

Panic stack:

```
panic: sync: negative WaitGroup counter

goroutine 45 [running]:
sync.(*WaitGroup).Add(0xc8202f8868, 0xffffffffffffffff)
        /usr/lib/go/src/sync/waitgroup.go:70 +0xd0
sync.(*WaitGroup).Done(0xc8202f8868)
        /usr/lib/go/src/sync/waitgroup.go:95 +0x2a
github.com/uber/tchannel-go.(*messageExchangeSet).removeExchange(0xc8202f8838, 0xc800000005)
        /home/udocker/moe/Godeps/_workspace/src/github.com/uber/tchannel-go/mex.go:202 +0x1f0
github.com/uber/tchannel-go.(*Connection).handleError(0xc8202f8700, 0xc8201c46c0)
        /home/udocker/moe/Godeps/_workspace/src/github.com/uber/tchannel-go/outbound.go:291 +0x685
github.com/uber/tchannel-go.(*Connection).readFrames(0xc8202f8700, 0x13)
        /home/udocker/moe/Godeps/_workspace/src/github.com/uber/tchannel-go/connection.go:636 +0x49d
created by github.com/uber/tchannel-go.(*Channel).newConnection
        /home/udocker/moe/Godeps/_workspace/src/github.com/uber/tchannel-go/connection.go:245 +0xa6f
```

--- a/connection.go
+++ b/connection.go
@@ -555,19 +555,20 @@ func (c *Connection) SendSystemError(id uint32, span *Span, err error) error {
 	// as we are sending the frame.
 	return c.withStateRLock(func() error {
 		// Errors cannot be sent if the connection has been closed.
-		if c.state != connectionClosed {
-			select {
-			case c.sendCh <- frame: // Good to go
-				return nil
-			default: // If the send buffer is full, log and return an error.
-			}
-			c.log.Warnf("Could not send error frame to %s for %d : %v",
-				c.remotePeerInfo, id, err)
-		} else {
+		if c.state == connectionClosed {
 			c.log.Infof("Could not send error frame to %s on closed conn for %d : %v",
 				c.remotePeerInfo, id, err)
+			return fmt.Errorf("failed to send error frame, connection state %v", c.state)
 		}
-		return fmt.Errorf("failed to send error frame")
+
+		select {
+		case c.sendCh <- frame: // Good to go
+			return nil
+		default: // If the send buffer is full, log and return an error.
+		}
+		c.log.Warnf("Could not send error frame to %s for %d : %v",
+			c.remotePeerInfo, id, err)
+		return fmt.Errorf("failed to send error frame, buffer full")
 	})
 }
 

--- a/connection_test.go
+++ b/connection_test.go
@@ -385,3 +385,17 @@ func TestWriteAfterTimeout(t *testing.T) {
 	})
 	VerifyNoBlockedGoroutines(t)
 }
+
+func TestReadTimeout(t *testing.T) {
+	WithVerifiedServer(t, nil, func(ch *Channel, hostPort string) {
+		for i := 0; i < 100; i++ {
+			ctx, cancel := NewContext(time.Second)
+			handler := func(ctx context.Context, args *raw.Args) (*raw.Res, error) {
+				defer cancel()
+				return nil, ErrTimeout
+			}
+			testutils.RegisterFunc(t, ch, "call", handler)
+			raw.Call(ctx, ch, hostPort, ch.PeerInfo().ServiceName, "call", nil, nil)
+		}
+	})
+}

--- a/mex.go
+++ b/mex.go
@@ -245,7 +245,7 @@ func (mexset *messageExchangeSet) forwardPeerFrame(frame *Frame) error {
 	}
 
 	if err := mex.forwardPeerFrame(frame); err != nil {
-		mexset.log.Warnf("Unable to forward frame ID %v type %v length %v to %s: %v",
+		mexset.log.Infof("Unable to forward frame ID %v type %v length %v to %s: %v",
 			frame.Header.ID, frame.Header.messageType, frame.Header.FrameSize(), mexset.name, err)
 		return err
 	}

--- a/outbound.go
+++ b/outbound.go
@@ -266,7 +266,7 @@ func (response *OutboundCallResponse) Arg3Reader() (io.ReadCloser, error) {
 	return response.arg3Reader()
 }
 
-// handleError andles an error coming back from the peer. If the error is a
+// handleError handles an error coming back from the peer. If the error is a
 // protocol level error, the entire connection will be closed.  If the error is
 // a request specific error, it will be written to the request's response
 // channel and converted into a SystemError returned from the next reader or

--- a/outbound.go
+++ b/outbound.go
@@ -289,7 +289,7 @@ func (c *Connection) handleError(frame *Frame) {
 	}
 
 	if err := c.outbound.forwardPeerFrame(frame); err != nil {
-		c.outbound.removeExchange(frame.Header.ID)
+		c.log.Infof("Failed to forward error frame %v to mex, error: %v", frame.Header, errMsg)
 	}
 }
 

